### PR TITLE
[FIX] fix arithmetic operation between numeric and non-numeric 

### PR DESCRIPTION
**Pull request by yangzhg**
fix arithmetic operation between numeric and non-numeric will cause unexpected value (#2142 #2145)

### CODE REVIEW

**Review by kangkaisen — DISMISSED** [review]
+1. LGTM

**Comment by kangkaisen** [comment]
@yangzhg Would better add a UT for this bug.

**Review by imay — CHANGES_REQUESTED** [review]
This change will make many already exist query fail, So I think it is not appropriate to change it directly.
Such as query `select 1 + "1"`

File: fe/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java [hunk 1]
```diff
@@ -133,6 +135,11 @@ public ArithmeticExpr(Operator op, Expr e1, Expr e2) {
         if (e2 != null) {
             children.add(e2);
         }
+        if (children.size() == 1) {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java]: Why generating sqlStr? I think it will make all objects generate an string, in some case that toSql is not called, sqlStr wouldn't be generated.

File: fe/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java [hunk 2]
```diff
@@ -133,6 +135,11 @@ public ArithmeticExpr(Operator op, Expr e1, Expr e2) {
...
+            sqlStr=op.toString() + " " + getChild(0).toSql();
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java]:
```suggestion
            sqlStr = op.toString() + " " + getChild(0).toSql();
```

File: be/src/exprs/cast_functions.cpp [hunk 3]
```diff
@@ -113,7 +112,9 @@ CAST_FUNCTION(FloatVal, DoubleVal, double_val)
         num_type ret; \
         ret.val = StringParser::string_parser_fn<native_type>( \
                 reinterpret_cast<char*>(val.ptr), val.len, &result); \
-        if (UNLIKELY(result != StringParser::PARSE_SUCCESS)) return num_type::null(); \
+        if (UNLIKELY(std::isnan(ret.val) || std::isinf(ret.val))) { \
+            ret.val = 0; \
```
Comment (imay) [inline be/src/exprs/cast_functions.cpp]: Should check if result is PARSE_SUCCESS

File: fe/src/main/java/org/apache/doris/analysis/StringLiteral.java [hunk 4]
```diff
@@ -139,12 +141,36 @@ public String getStringValue() {
 
     @Override
     public long getLongValue() {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/analysis/StringLiteral.java]: Better to add some UT to test corner case such as "+abc", "-.a"

File: fe/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java (hunk 1 above)
Comment (yangzhg) [reply fe/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java]: the expr after analyze will be changed, this will cause output table header and user input sql are different e.g. 
image
generating sqlStr at constructor will keep the origin sql 
image

**Comment by kangkaisen** [comment]
I test with mysql 5.7.1. 
I think we should keep consistent with mysql.  
We could compute the following arithmetic  expression In FE, and we should add UT in FE to cover this.

```
mysql> select 1 + "kks";
+-----------+
| 1 + "kks" |
+-----------+
|         1 |
+-----------+
1 row in set, 1 warning (0.00 sec)

mysql> select 1 - "kks";
+-----------+
| 1 - "kks" |
+-----------+
|         1 |
+-----------+
1 row in set, 1 warning (0.05 sec)

mysql> select 1 * "kks";
+-----------+
| 1 * "kks" |
+-----------+
|         0 |
+-----------+
1 row in set, 1 warning (0.00 sec)

mysql> select 1 /  "kks";
+------------+
| 1 /  "kks" |
+------------+
|       NULL |
+------------+
1 row in set, 1 warning (0.00 sec)
```

**Comment by yangzhg** [comment]
> I test with mysql 5.7.1.
> I think we should keep consistent with mysql.
> We could compute the following arithmetic expression In FE, and we should add UT in FE to cover this.
> 
> ```
> mysql> select 1 + "kks";
> +-----------+
> | 1 + "kks" |
> +-----------+
> |         1 |
> +-----------+
> 1 row in set, 1 warning (0.00 sec)
> 
> mysql> select 1 - "kks";
> +-----------+
> | 1 - "kks" |
> +-----------+
> |         1 |
> +-----------+
> 1 row in set, 1 warning (0.05 sec)
> 
> mysql> select 1 * "kks";
> +-----------+
> | 1 * "kks" |
> +-----------+
> |         0 |
> +-----------+
> 1 row in set, 1 warning (0.00 sec)
> 
> mysql> select 1 /  "kks";
> +------------+
> | 1 /  "kks" |
> +------------+
> |       NULL |
> +------------+
> 1 row in set, 1 warning (0.00 sec)
> ```

i have done this 
the current output :
```
mysql> select 1 +  "kks";
+-----------+
| 1 + 'kks' |
+-----------+
|         1 |
+-----------+
1 row in set (0.02 sec)

mysql> select 1 -  "kks";
+-----------+
| 1 - 'kks' |
+-----------+
|         1 |
+-----------+
1 row in set (0.01 sec)

mysql> select 1 *  "kks";
+-----------+
| 1 * 'kks' |
+-----------+
|         0 |
+-----------+
1 row in set (0.01 sec)

mysql> select 1 /  "kks";
+-----------+
| 1 / 'kks' |
+-----------+
|      NULL |
+-----------+
1 row in set (0.01 sec)
```

File: fe/src/main/java/org/apache/doris/analysis/StringLiteral.java [hunk 5]
```diff
@@ -169,6 +195,75 @@ private LiteralExpr convertToDate(Type targetType) throws AnalysisException {
...
+     * 123a -> 123
+     * -123 -> -123
+     * 123.123 -> 123.123
+     * 123e4 -> 123e4
+     * 123e-4 -> 123e-4
+     * @param str
+     * @return
+     */
+    protected static String leadingNum(String str) {
+ String in = str.trim();
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/analysis/StringLiteral.java]:
```suggestion
        String in = str.trim();
```

Comment (yangzhg) [reply fe/src/main/java/org/apache/doris/analysis/StringLiteral.java]: done

**Review by imay — APPROVED** [review]
LGTM
